04-Protocol-Security.md

```diff
@@ -0,0 +1,392 @@
+
+#### 4.5.5.1 Signature structure
+Schnorr signature with *key prefixing* is used<sup>[3](#reference-3)</sup>
+
+signature is constructed for
+* message `m`, where `m` is `HASH` of the serialized fields of the `CERTIFICATE` that are marked for signing, i.e. `m = SHA-256(version || valid_from || not_valid_after || server_public_key)`
+* public key `P` that is Certificate Authority
+
+Signature itself is concatenation of an EC point `R` and an integer `s` (note that each item is serialized as 32 bytes array) for which identity `s⋅G = R + HASH(R || P || m)⋅P` holds.
+
+### 4.5.6 Transport message encryption and format
+
+After handshake process is finished, both initiator and responder have CipherState objects for encryption and decryption and after initiator validated server's identity, any subsequent traffic is encrypted and decrypted with `EncryptWithAd()` and `DecryptWithAd()` methods of the respectrive CipherState objects with zero-length associated data.
+
+Ciphertext is sent in `NOISE_FRAME` over the wire.
+```
+NOISE_FRAME
++-------------+-----------+--------------------------------------------------------------------------------------------+
+| Field Name  | Data Type | Description                                                                                |
++-------------+-----------+--------------------------------------------------------------------------------------------+
+| ciphertext  | B0_64K    | AEAD ciphertext including 16 Bytes MAC                                                     |
++-------------+-----------+--------------------------------------------------------------------------------------------+
+Message length: <Plaintext length> + 18 bytes = <Plaintext length> + <MAC length> + <Type length prefix>
+
+Maximum message length = 65537 bytes
+Maximum ciphertext length = 65535 bytes
+Maximum plaintext length 65519 bytes
+```
+
+
+Note that in regard to Stratum V2 message, `NOISE_FRAME` doesn't necessarily need to contain to exactly one encrypted Stratum message. Ciphertext payload may contain multiple subsequent messages or even only partial message. Examples:
+* `OpenStandardMiningChannelSuccess` followed immediately with `NewMiningJob`
+* Arbitrary message containing `B0_16M` type, since the noise ciphertext can be at most `2**16 - 1 == 65535` bytes long
+
+
+## 4.6 URL Scheme and Pool Authority Key
+Downstream nodes that want to use the above outlined security scheme need to have configured the **Pool Authority Public Key** of the pool that they intend to connect to. It is provided by the target pool and communicated to its users via a trusted channel.
+At least, it can be published on the pool's public website.
+
+
+The key can be embedded into the mining URL as part of the path.
+
+Authority Public key is encoded as a 32-byte secp256k1 public key (with implicit Y coordinate), prefixed with `[0x4b, 0x69]`, in [base58-check](https://en.bitcoin.it/wiki/Base58Check_encoding) encoding.
```

Comment:
Maybe we don't have to use exactly this prefix and leave it undefined. Individual users can use these bytes to make the base58 encoded public key start with string prefix of their choice.